homeassistant/components/lyric/__init__.py

```diff
@@ -0,0 +1,207 @@
+"""The Honeywell Lyric integration."""
+import asyncio
+from datetime import timedelta
+import logging
+from typing import Any, Dict, List
+
+from aiohttp.client_exceptions import ClientResponseError
+from aiolyric import Lyric
+from aiolyric.exceptions import LyricAuthenticationException, LyricException
+from aiolyric.objects.device import LyricDevice
+from aiolyric.objects.location import LyricLocation
+import async_timeout
+import voluptuous as vol
+
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.const import CONF_CLIENT_ID, CONF_CLIENT_SECRET
+from homeassistant.core import HomeAssistant
+from homeassistant.exceptions import ConfigEntryNotReady
+from homeassistant.helpers import (
+    aiohttp_client,
+    config_entry_oauth2_flow,
+    config_validation as cv,
+)
+from homeassistant.helpers.update_coordinator import (
+    CoordinatorEntity,
+    DataUpdateCoordinator,
+    UpdateFailed,
+)
+
+from .api import ConfigEntryLyricClient, LyricLocalOAuth2Implementation
+from .config_flow import OAuth2FlowHandler
+from .const import DATA_COORDINATOR, DATA_LYRIC, DOMAIN, OAUTH2_AUTHORIZE, OAUTH2_TOKEN
+
+CONFIG_SCHEMA = vol.Schema(
+    {
+        DOMAIN: vol.Schema(
+            {
+                vol.Required(CONF_CLIENT_ID): cv.string,
+                vol.Required(CONF_CLIENT_SECRET): cv.string,
+            }
+        )
+    },
+    extra=vol.ALLOW_EXTRA,
+)
+
+_LOGGER = logging.getLogger(__name__)
+
+PLATFORMS = ["climate", "sensor"]
+
+
+async def async_setup(hass: HomeAssistant, config: dict):
+    """Set up the Honeywell Lyric component."""
+    hass.data[DOMAIN] = {}
+
+    if DOMAIN not in config:
+        return True
+
+    hass.data[DOMAIN][CONF_CLIENT_ID] = config[DOMAIN][CONF_CLIENT_ID]
+
+    OAuth2FlowHandler.async_register_implementation(
+        hass,
+        LyricLocalOAuth2Implementation(
+            hass,
+            DOMAIN,
+            config[DOMAIN][CONF_CLIENT_ID],
+            config[DOMAIN][CONF_CLIENT_SECRET],
+            OAUTH2_AUTHORIZE,
+            OAUTH2_TOKEN,
+        ),
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up Honeywell Lyric from a config entry."""
+    implementation = (
+        await config_entry_oauth2_flow.async_get_config_entry_implementation(
+            hass, entry
+        )
+    )
+
+    session = aiohttp_client.async_get_clientsession(hass)
+    oauth_session = config_entry_oauth2_flow.OAuth2Session(hass, entry, implementation)
+
+    client = ConfigEntryLyricClient(session, oauth_session)
+
+    client_id = hass.data[DOMAIN][CONF_CLIENT_ID]
+    lyric = Lyric(client, client_id)
+
+    try:
+        await lyric.get_locations()
+    except (
+        LyricAuthenticationException,
+        LyricException,
+        ClientResponseError,
+    ) as exception:
+        _LOGGER.warning(exception)
+        raise ConfigEntryNotReady from exception
```

Comment:
Need to add reauthentication here. Honeywell have a tendency to clear tokens.



Maybe this can be done as a part of the oauth implementation? Anyone know if this is possible?